src/udev/dmi_memory_id/dmi_memory_id.c

```diff
@@ -0,0 +1,806 @@
+                if (h.type == 127 && (flags & FLAG_STOP_AT_EOT))
+                        break;
+        }
+}
+
+static void dmi_table(off_t base, uint32_t len, uint16_t num, const char *devmem, uint32_t flags) {
+        _cleanup_free_ uint8_t *buf = NULL;
+        size_t size;
+        int r;
+
+        /*
+         * When reading from sysfs or from a dump file, the file may be
+         * shorter than announced. For SMBIOS v3 this is expcted, as we
+         * only know the maximum table size, not the actual table size.
+         * For older implementations (and for SMBIOS v3 too), this
+         * would be the result of the kernel truncating the table on
+         * parse error.
+         */
+        r = read_full_file_full(AT_FDCWD, devmem,
+                        flags & FLAG_NO_FILE_OFFSET ? 0 : base, len,
+                        0, NULL, (char **) &buf, &size);
+        if (r < 0) {
+                log_error_errno(r, "Failed to read table, sorry: %m");
+                return;
+        }
+
+        len = size;
+        dmi_table_decode(buf, len, num, flags);
+}
+
+/* Same thing for SMBIOS3 entry points */
+static int smbios3_decode(uint8_t *buf, const char *devmem, uint32_t flags) {
+        ULargeInteger offset;
+
+        /* Don't let checksum run beyond the buffer */
+        if (buf[0x06] > 0x20) {
+                log_error("Entry point length too large (%u bytes, expected %u).",
+                          (unsigned)buf[0x06], 0x18U);
+                return 0;
+        }
+
+        if (!checksum(buf, buf[0x06]))
+                return 0;
+
+        offset = QWORD(buf + 0x10);
+        if (!(flags & FLAG_NO_FILE_OFFSET) && offset.h && sizeof(off_t) < 8) {
+                log_error("64-bit addresses not supported, sorry.");
+                return 0;
+        }
+
+        dmi_table(((off_t)offset.h << 32) | offset.l,
+                        DWORD(buf + 0x0C), 0, devmem, flags | FLAG_STOP_AT_EOT);
+
+        return 1;
+}
+
+static int smbios_decode(uint8_t *buf, const char *devmem, uint32_t flags) {
+        /* Don't let checksum run beyond the buffer */
+        if (buf[0x05] > 0x20) {
+                log_error("Entry point length too large (%u bytes, expected %u).",
+                          (unsigned)buf[0x05], 0x1FU);
+                return 0;
+        }
+
+        if (!checksum(buf, buf[0x05])
+            || memcmp(buf + 0x10, "_DMI_", 5) != 0
+            || !checksum(buf + 0x10, 0x0F))
+                return 0;
+
+        dmi_table(DWORD(buf + 0x18), WORD(buf + 0x16), WORD(buf + 0x1C),
+                  devmem, flags);
+
+        return 1;
+}
+
+static int legacy_decode(uint8_t *buf, const char *devmem, uint32_t flags) {
+        if (!checksum(buf, 0x0F))
+                return 0;
+
+        dmi_table(DWORD(buf + 0x08), WORD(buf + 0x06), WORD(buf + 0x0C),
+                  devmem, flags);
+
+        return 1;
+}
+
+int main(int argc, char * const argv[]) {
```

Comment:
Please do
```C
static int run(int argc, const char** argv) {
       ...
}
DEFINE_MAIN_FUNC(run);
```
and use normal negative returns in `run`.

Reply:
``DEFINE_MAIN_FUNCTION(run);```